File: FsCheck.Documentation/TestData.fs
module TestData

open FsCheck

let chooseFromList xs = 
    gen { let! i = Gen.choose (0, List.length xs-1) 
          return (List.nth xs i) }
          
let chooseBool = 
    Gen.oneof [ gen { return true }; gen { return false } ]

let chooseBoolFrequency = 
    Gen.frequency [ (2, gen { return true }); (1, gen { return false })]

//The size of test data
let matrix gen = Gen.sized <| fun s -> Gen.resize (s|>float|>sqrt|>int) gen

//Generating Recusrive data types
type Tree = Leaf of int | Branch of Tree * Tree

let rec unsafeTree() = 
    Gen.oneof [ Gen.map Leaf Arb.generate<int> 
                Gen.map2 (fun x y -> Branch (x,y)) (unsafeTree()) (unsafeTree())]

let tree =
    let rec tree' s = 
        match s with
        | 0 -> Gen.map Leaf Arb.generate<int>
        | n when n>0 -> 
            let subtree = tree' (n/2)
            Gen.oneof [ Gen.map Leaf Arb.generate<int> 
                        Gen.map2 (fun x y -> Branch (x,y)) subtree subtree]
        | _ -> invalidArg "s" "Only positive arguments are allowed"
    Gen.sized tree'
  
type Box<'a> = Whitebox of 'a | Blackbox of 'a

let boxGen<'a> : Gen<Box<'a>> = 
    gen { let! a = Arb.generate
          return! Gen.elements [ Whitebox a; Blackbox a] }
    
type MyGenerators =
    static member Tree() =
        {new Arbitrary<Tree>() with
            override x.Generator = tree
            override x.Shrinker t = Seq.empty }
    static member Box() = Arb.fromGen boxGen
Arb.register<MyGenerators>() |> ignore

let RevRevTree (xs:list<Tree>) = 
    List.rev(List.rev xs) = xs
Check.Quick RevRevTree

let RevRevBox (xs:list<Box<int>>) = 
    List.rev(List.rev xs) = xs
    |> Prop.collect xs
Check.Quick RevRevBox

File: examples/CSharp.DocSnippets/TestData.cs
using FsCheck;
using FsCheck.Fluent;
using System;

namespace CSharp.DocSnippets {

    public abstract class Tree {

    }

    public class Leaf : Tree {
        public int Value { get; private set; }

        public Leaf(int value) {
            Value = value;
        }
    }

    public class Branch : Tree {
        public Tree Left { get; private set; }
        public Tree Right { get; private set; }

        public Branch(Tree left, Tree right) {
            Left = left;
            Right = right;
        }
    }

    class TestData {

        //[chooseFrom]
        public Gen<T> ChooseFrom<T>(T[] xs) {
            return from i in Gen.Choose(0,xs.Length-1)
                   select xs[i];
        }
        //[/chooseFrom]

        //[matrixGen]
        public static Gen<T> Matrix<T>(Gen<T> generator) {
            return Gen.Sized(s => generator.Resize(Convert.ToInt32(Math.Sqrt(s))));
        }
        //[/matrixGen]

        //[unsafeTree]
        public static Gen<Tree> UnsafeTree() {
            return Gen.OneOf(ArbMap.Default.GeneratorFor<int>().Select(i => (Tree) new Leaf(i)),
                             UnsafeTree().Two().Select(t => (Tree) new Branch(t.Item1,t.Item2)));
        }
        //[/unsafeTree]

        //[safeTree]
        public static Gen<Tree> SafeTreeHelper(int size) {
            if (size == 0) {
                return ArbMap.Default.GeneratorFor<int>().Select(i => (Tree)new Leaf(i));
            }
            else {
                var subtree = SafeTreeHelper(size / 2);
                return Gen.OneOf(ArbMap.Default.GeneratorFor<int>().Select(i => (Tree) new Leaf(i)),
                                 subtree.Two().Select(t => (Tree) new Branch(t.Item1,t.Item2)));
            }
        }

        public static Gen<Tree> SafeTree() {
            return Gen.Sized(SafeTreeHelper);
        }
        //[/safeTree]

        //[MyGenerators]
        public class MyGenerators {
            public static Arbitrary<Tree> Trees() {
                return SafeTree().ToArbitrary();
            }
        }
        //[/MyGenerators]

        public static void Samples() {
            //[chooseBool]
            var chooseBool = Gen.OneOf(Gen.Constant(true), Gen.Constant(false));
            //[/chooseBool]

            //[chooseBool2]
            var chooseBool2 = Gen.Frequency(
                (2, Gen.Constant(true)),
                (1, Gen.Constant(false)));
            //[/chooseBool2]

            //[sizedInt]
            var sizedInt = Gen.Sized(s => Gen.Choose(0, s));
            //[/sizedInt]

            //[register]
            Check.One(Config.Default.WithArbitrary(new[] { typeof(MyGenerators) }), true);
            //[/register]
        }

    }
}
